Feat/teams component

manvinderjit opened the pull request:
Add Teams Component to show TorontoJS' teams.
Add TeamMember Component to show team members.
Updated index.css file to fix open props import and variables.

Resolves: #8

manvinderjit committed:
fix: Fix open-props import and vars

Fix open props import to use npm package.
Fix variable declaration for root variables in index.css

manvinderjit committed:
feat: Add Teams Component

Add Teams Component to show TorontoJS' teams.
Add TeamMember Component to show team members.

madcampos commented on public/members.json:
```diff
@@ -0,0 +1,105 @@
+    "description": "A passionate volunteer.",
+    "happenedAt": "2024-12-09T12:00:00",
+    "insertedAt": "2024-12-01T10:00:00",
+    "avatar": "https://randomuser.me/api/portraits/men/1.jpg",
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** being super absurdly picky here, can we have some variation on the data, specially the avatars? 🙃

what about using [faker](https://fakerjs.dev/) to generate us some data?‏

manvinderjit replied:
Resolved: Added distinct names and avatars images for team members

madcampos commented on public/teams.json:
```diff
@@ -0,0 +1,34 @@
+    },
+    {
+        "id": "ae2379dc-f215-4c79-a942-45ccb308312a",
+        "name": "Omnipotent Organisers",
```
<!-- Praises highlight something positive. Try to leave at least one of these comments per review. Do not leave false praise (which can actually be damaging). Do look for something to sincerely praise.  -->
**praise:** Love the team names! 💖‏

madcampos commented on src/components/TeamMember/TeamMember.jsx:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** ‏I think it would be better to call this component "team member card" or "member card". It makes the name more explicit on what it is and how it is used.

The reasoning is: imagine you open a link to this file on github, without any other context, is it easy to figure out what this component do or how it relate to the application as a whole?

madcampos commented on src/components/TeamMember/TeamMember.jsx:
```diff
@@ -0,0 +1,21 @@
+const TeamMember = ({ member }) => {
+  return (
+    <div className="team-member-card">
+      <a href={member.socialLinks.linkedin} target="_blank">
+        <div className="team-member-profile">
+          <img
+            className="avatar"
+            src={member.avatar || "/default-avatar.png"}
+            alt={`Team ${member.name} Avatar`}
+          />
+          <div className="team-member-card-text">
+            <h3 className="team-member-card-heading">{member.name}</h3>
+            <span>linkedin</span>
+          </div>
+        </div>
+      </a>
+    </div>
```
```suggestion
    <article className="team-member-card">
      <a
        href={member.socialLinks.linkedin}
        target="_blank"
        className="team-member-profile"
      >
        <picture>
          <img
            className="avatar"
            src={member.avatar || "/default-avatar.png"}
            alt={`Team ${member.name} Avatar`}
          />
        </picture>
        <header className="team-member-card-text">
          <h3 className="team-member-card-heading">{member.name}</h3>
        </header>
        <div className="social-media-links">
          <ul aria-label="Social Media Links">
            <li><a>linkedin</a></li>
          </ul>
        </div>
      </a>
    </article>
```

**Explanation:**

- Use semantic tags, those relate to [landmark roles](https://www.w3.org/WAI/ARIA/apg/practices/landmark-regions/) and make navigation easier.
  - Specifically, `<article>` here means "an item on a list of items". It is not necessarily a "news _article_", it's meaning is more like an "_article_ of clothing" (i.e. an item on a shopping website). That makes the tag a good candidate for a "card".
  - `<header>` is also a landmark role and helps with organizing content inside of an `<article>`. Other tags that help are `<footer>` and `<aside>`.
- The `<div>` for social media links is just to help partition content, it has no special meaning and doesn't affect the overall structure, so it is okay to use a `<div>` here
- The list however, _does_ have a meaning.
  - Here we want to add a list of links, to having a `<ul>` conveys that.
  - When a screen reader user navigates to the list, the `aria-label` acts as an extra [affordance](https://en.wikipedia.org/wiki/Affordance) to help contextualize the user where they are in the page and what the expected content is.
  - For sighted users, the affordance[^1] here will be icons and text for the social media links.
- Lastly, wrapping the `<img>` in a `<picture>` element helps with styling the image and giving a container to do whatever we want without removing semantic meaning.

Here is an extra reference, it doesn't include the landmark elements, but is a good read on cards and accessibility, albeit a little bit old: https://inclusive-components.design/cards/

[^1]: Cool video about UX but somewhat unrelated to this: https://youtu.be/yY96hTb8WgI?si=dELfNYcuG7GaQKgt

manvinderjit replied:
For now, I was following the semantics on the current volunteer page and the w3c design, but will adapt these changes from now on.

madcampos commented on src/App.jsx:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** Let's use this PR as a base: https://github.com/torontojs/volunteer/pull/12

It changes a little bit the structure to add two different pages as separate "apps", with many quotes here. One for profile and other for team.

That way we can build them as separate pages and not worry about react router or other shenanigans like that, making the app closer to a multi-page application than a single page application‏

manvinderjit replied:
Do you want me to clone it right away or wait for the pull request to be merged?

madcampos replied:
Let's merge [that PR](https://github.com/torontojs/volunteer/pull/12) first and then you can rebase/merge on top of this.

I'm waiting for @kbventures review, but please feel free to also chime in on the PR

madcampos commented on src/App.css:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** I'd rather we have a separate css file for each component with all the styles relative to it there.

To make things easier to load, we can add a `<link>` tag for each component css or have a `main.css` file that `@import` all of the components styles.

manvinderjit replied:
Separated css files for components

madcampos commented on src/index.css:
```diff
@@ -1,6 +1,10 @@
-@import url('https://cdn.jsdelivr.net/npm/open-props@1.0.0/open-props.min.css');
+@import "open-props/style";
+@import "open-props/open-props.min.css";
```
```suggestion
@import "open-props/open-props.min.css";
```

This is a redundant import, let's use the direct css file one.

madcampos commented on src/components/Teams/Teams.jsx:
```diff
@@ -0,0 +1,85 @@
+
+  useEffect(() => {
+  
+    //TODO: Add abort controller when fetching from API
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** Why would we need an abort controller here?‏

manvinderjit replied:
In case we want to abort the request to API if the component reloads without the previous request being fulfilled. It is not a necessity though.

madcampos replied:
You mean so re-rendering the component doesn't request the data twice?

To solve this problem, you can have two bits of information:
1. A flag telling if a request is in progress.
2. The data itself, it can be an object or nullish.

Then the `useEffect` can look like this:
1. Check if there is a request in progress and just `return`.
2. Check if the data exists (i.e. is not `null`/`undefined`) and just `return`.
3. Add the logic for the data request itself, taking care to assign both the data and the flag.

This way even if the component re-renders it will not send multiple requests. 😉

madcampos commented on src/components/Teams/Teams.jsx:
```diff
@@ -0,0 +1,85 @@
+    // Fetch data from the JSON file in the public directory
+    async function fetchData () {
+      try {
+        const [responseTeams, responseMembers] =  await Promise.all([fetch("/teams.json"), fetch("/members.json")]);
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** Separate the two requests so they are independent.

We should first load the team data and then only if that loads we load the members data. It is okay to display only the teams information and not the members.

madcampos commented on src/components/Teams/Teams.jsx:
```diff
@@ -0,0 +1,85 @@
+
+  // If data not yet loaded
+  if (!isLoadedTeamsData) {
+    return <div>Loading...</div>;
```
```suggestion
    return <div aria-live="polite" role="status">Loading teams...</div>;
```

Here is the reference for [ARIA Live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions).

The TLDR:
- Live region will announce what is going on to screen readers, we use the `polite` value here as it will not interrupt what the screen reader is reading. As the message about loading data is not urgent it is better to wait and not interrupt things.
- The `role="status"` identifies this `<div>` as a place where the status of the page is present.
- The added text helps give context of that is happening.

madcampos commented on src/components/Teams/Teams.jsx:
```diff
@@ -0,0 +1,85 @@
+
+  // If error encountered
+  if (teamsError) {
+    return <div>An error occurred!</div>;
```
```suggestion
    return <div>Unable to load teams. Please try refreshing the page.</div>;
```

As a UX improvement we give a more detailed error message, explaining what happened and also provide a potential fix for it.

[This heuristic](https://www.nngroup.com/articles/ten-usability-heuristics/#toc-9-help-users-recognize-diagnose-and-recover-from-errors-9) for UX design in specific should be present here.

madcampos commented on src/components/Teams/Teams.jsx:
```diff
@@ -0,0 +1,85 @@
+
+  // If team data is loaded but empty data returned
+  if (isLoadedTeamsData && !teamsError && (!teamsData || teamsData.length === 0)) {
+    return <div>No data found!</div>;
```
Again here: provide a better error message and how to recover from that.

madcampos commented on src/components/Teams/Teams.jsx:
```diff
@@ -0,0 +1,85 @@
+  // Render data
+  if(!teamsError && teamsData.length > 0) {
+    return (
+      <section>
+        {teamsData.map((team, index) => {
+          return (
+            <div key={index} className="team">
+              <div className="team-header">
+                <h2>{team.name}</h2>
+                <p>{team.description}</p>
+              </div>
+
+              <div className="team-body">              
+                {(!teamMembersData || teamMembersData.length <= 0) && (
+                  <div>No Team Members Found!</div>
+                )}                
+                {teamMembersData.length > 0 &&
+                  teamMembersData.map((member, index) => (
+                    <TeamMember key={index} member={member} />
+                  ))}
+              </div>
+            </div>
+          );
+        })}
+      </section>
```
```suggestion
      <main>
        {teamsData.map((team, index) => {
          return (
            <article key={index} className="team">
              <header className="team-header">
                <h2>{team.name}</h2>
              </header>

              <div className="team-body">
                <p>{team.description}</p>
         
                {(!teamMembersData || teamMembersData.length <= 0) && (
                  <div>This team has no active members.</div>
                )}                
                {teamMembersData.length > 0 && (
                  <h3 id="team-members-label-{index}">Team members</h3>
                  <ul aria-labeled-by="team-members-label-{index}">
                  {teamMembersData.map((member, index) => (
                    <li><TeamMember key={index} member={member} /></li>
                  ))}
                  </ul>
                )}
              </div>
            </article>
          );
        })}
      </main>
```

Explanation:

- `<main>` represents the main content on the page, if we are rendering only this component, it should render as the main element. Please note there should only be **ONE** `<main>` element on each page.
- The team description can be part of the team body.
- The team members heading helps with content structure.
- The team members should be a list, it also has a label defined using the `aria-labeledby` attribute. The attribute has to reference an `id`, because of that, the id is constructed using the index of each team.
- Again, `<header>` and `<article>` are used as landmarks.

madcampos commented on src/App.css:
```diff
@@ -42,3 +42,127 @@
+  position: relative;  
+}
+
+.team-member-card span::after {  
+  content: '';
+  position: absolute;
+  transition: width .3s ease-in-out;
+  bottom: 0;
+  left: 50%;
+  width: 0;
+  height: 1px;
+  background-color:var(--color-grey);  
+  transform: translateX(-50%);
+}
+
+.team-member-card:hover span::after {
+  width: 100%;
+  left: 50%;
+}
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** Transitioning `width` and `height` force the browser to do expensive layout and content flow calculations‏, even if the element doesn't affect the parent.

As the intended effect is to change the background color, using a `linear-gradient` and changing the `background-size` property will work, as well as the `transform` property with a `scale` value.

Those properties will only affect the paint process and are GPU accelerated.

Here is a reference of which property triggers what: https://css-triggers.com/

In order of cheap to expensive: paint > layout > composite

madcampos commented on src/components/Teams/Teams.jsx:
```diff
@@ -0,0 +1,85 @@
+                )}                
+                {teamMembersData.length > 0 &&
+                  teamMembersData.map((member, index) => (
+                    <TeamMember key={index} member={member} />
```
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought:** I don't think we want to load the team members on the team list page... It makes more sense to have a page listing teams have only the names and a short summary of the team, not much information there, and if the person want to see more, they go to the team page itself where all information is listed.

@kbventures @marianavrocha thought?

manvinderjit committed:
Update src/components/TeamMember/TeamMember.jsx

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
Update src/index.css

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
Update src/components/Teams/Teams.jsx

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
Update src/components/Teams/Teams.jsx

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
Update src/components/Teams/Teams.jsx

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
fix: Apply multiple fixes to Teams component

Rename TeamMember component to TeamMemberCard.
Separate CSS files for each component.
Separate API requests for teams and team members.
Update Error Messages.
Update transitions to replace layout-intensive effects.
Move Profile component to its own directory.

manvinderjit committed:
Fix: Resolve conflicts after Team Component fixes

Following fixes were applied in commit #41479847
Rename TeamMember component to TeamMemberCard.
Separate CSS files for each component.
Separate API requests for teams and team members.
Update Error Messages.
Update transitions to replace layout-intensive effects.
Move Profile component to its own directory.

10 files changed, 330 insertions(+), 309 deletions(-)

manvinderjit commented:
@madcampos Resolved all except the following:
1.  https://github.com/torontojs/volunteer/pull/13#discussion_r1888851747
Waiting on PR merge for https://github.com/torontojs/volunteer/pull/12 
2. [Flags for reload of useEffect](https://github.com/torontojs/volunteer/pull/13#discussion_r1888869002)
Will add when fetching data from API
3. Separating Team and Team Members pages https://github.com/torontojs/volunteer/pull/13#discussion_r1888926676
Waiting on confirmation

manvinderjit committed:
fix: Add aria live regions for status updates

Add aria live regions for status updates in Teams Component.

madcampos commented on src/components/Profile/Profile.jsx:
Here we can use a similar structure as the team to create a card using `<article>` and friends.

The same patterns should be used.

manvinderjit replied:
It was a component created by Ken when he setup the project. I removed it from the component tree in App.js for the moment but didn't delete the file. I don't know if it is to be included or not.

madcampos replied:
I see, I thought it was the component used for displaying profiles.

We do want to display profiles and use the semantic html pattern for it. But if it is using this component or not, it is up to you 😉

manvinderjit commented:
I will give merging a shot and in case I have any issues, I will ping on slack.

manvinderjit committed:
merge: Merge with dev branch
Merge with dev branch at commit #2c3e5ba8.
Update structure in accordance with the dev branch.

manvinderjit commented:
@madcampos I have merged the two branches and pushed. Howver, there seems to be a CORS issue with the deployment. Please do look into it.

manvinderjit committed:
fix: Fix config

Fix base url.
Add netlify config file.

manvinderjit committed:
fix: Remove base URL

manvinderjit committed:
cleanup: Cleanup code vite config

Cleanup vite config to remove base URL.
